### ISSUE TITLE
Unexpected non-string and non-number key while sanitizing a Map-like type

### ISSUE THREAD

**Issue by hqwangningbo** [paritytech/substrate-api-sidecar#1082]
**Description**

substrate-spi-sidecar lastest version
wss: wss://bifrost-rpc.liebi.com/ws

The query of other blocks is no problem
Querying this block reports an error

<img width="1427" alt="image" src="https://user-images.githubusercontent.com/57781136/195234034-d85e2128-0453-42e5-9cd3-043dd8d3bbee.png">

The block information found in subscan is as follows:
https://bifrost-kusama.subscan.io/block/2652989

**Comment by hqwangningbo** [comment]
mark

**Comment by TarikGul** [comment]
Hey @hqwangningbo, sorry for the delayed response as I am finally back from some time off. I was able to reproduce the following, and will start debugging this. I looked at the general data in the block and couldn't find any irregularities for sanitization from first look but will keep you updated.

**Comment by TarikGul** [comment]
So the type its failing on is: `{"parents":"u8","interior":"XcmV1MultilocationJunctions"}`

**Comment by hqwangningbo** [comment]
Okay, I get it, will you fix it here?

**Comment by TarikGul** [comment]
Yea looking for a solution today. Its a little tricky since its a large nested type, and its all done recursively, but Ill have something put together

**Comment by TarikGul** [comment]
Here is the reproduction in polkadot-js:

```
require('@polkadot/api-augment');
const { ApiPromise, WsProvider } = require('@polkadot/api');

/**
 * Simple polkadot-js template to test and reproduce things locally
 */
const main = async () => {
    const api = await ApiPromise.create({
        provider: new WsProvider('wss://bifrost-rpc.liebi.com/ws'),
    });

    const block = await api.rpc.chain.getBlock('0xc860167557b095b858aba3940d90b76b2ed739122c10737569af9b2979d5960c');
    
    // asMulti method extrinsic
    const ext = block.block.extrinsics[2];
    // Call we are going to decode and look into
    const callArg = ext.method.args[3];
    const call = api.registry.createType('Call', callArg.toHex());

    const arg = call.args[0][0]
    
    console.log('[Call Decoded]: ', call.args[0][0])

    /**
     * {
            moonbeam: {
                account: { parents: 1, interior: [Object] },
                delegations: {
                    '{"parents":1,"interior":{"x2":[{"parachain":2023},{"accountKey20":{"network":{"any":null},"key":"0xc1b883f2c1d6dcd0672062b359a754dd100cfbea"}}]}}': '0x000000000000000046c94f94916881ee'
                },
                total: '0x000000000000000046c94f94916881ee',
                lessTotal: 0,
                requests: [],
                requestBriefs: {},
                status: { active: null }
            }
        }
     */
    console.log(arg.args[2].toJSON())

    // Option<BifrostSlpPrimitivesLedger>
    console.log(arg.args[2].toRawType())
};

main().finally(() => process.exit());
```

**Comment by TarikGul** [comment]
@hqwangningbo This seems to be more of an issue with the types given from bifrost-kusama. Not much I can do.

I tried looking for the following type through your repos (`Option<BifrostSlpPrimitivesLedger>`), and couldn't find anything.

**Comment by TarikGul** [comment]
Maybe this is a `polkadot-js` issue?

**Comment by Imod7** [comment]
The same error was also reported from another team : 
- while connected to Bifrost 
- with Sidecar `v14.2.0` but also latest version `v14.3.1`
- at block height 3460065 (`blocks/3460065`)
- error
```
{
  "code": 500,
  "message": "Unexpected non-string and non-number key while sanitizing a Map-like type",
  "stack": "InternalServerError: Unexpected non-string and non-number key while sanitizing a Map-like type\n    at /..../substrate-api-sidecar/build/src/sanitize/sanitizeNumbers.js:174:19\n    at Type.forEach (<anonymous>)\n    at mapTypeSanitizeKeyValue (/.../substrate-api-sidecar/build/src/sanitize/sanitizeNumbers.js:171:9)\n    at sanitizeCodec (/.../substrate-api-sidecar/build/src/sanitize/sanitizeNumbers.js:92:16)\n    at sanitizeNumbers (/.../substrate-api-sidecar/build/src/sanitize/sanitizeNumbers.js:125:16)\n    at /.../substrate-api-sidecar/build/src/sanitize/sanitizeNumbers.js:50:31\n    at Array.reduce (<anonymous>)\n    at sanitizeCodec (/.../substrate-api-sidecar/build/src/sanitize/sanitizeNumbers.js:45:30)\n    at sanitizeNumbers (/.../substrate-api-sidecar/build/src/sanitize/sanitizeNumbers.js:125:16)\n    at sanitizeCodec (/.../substrate-api-sidecar/build/src/sanitize/sanitizeNumbers.js:76:56)",
  "level": "error"
}
```

Since this issue is still open, I wonder if this error is still up for investigation or it is bifrost specific so we _should not_ (as per _we should keep Parity supported Sidecar as generic as possible_) provide support for that ?

**Comment by TarikGul** [comment]
@hqwangningbo Any idea on how we can fix the above?

Seems like users who use sidecar with bifrost are also experiencing this issue.

**Comment by TarikGul** [comment]
Pinging @herryho to bump this forward a bit.

**Comment by web3olalala** [comment]
@TarikGul Our Key is an Object, not string or number, so report this error, you can see the log of his Key corresponds to the type of Multilcation, so I want to Object stringify into a string and then go to deal with it, I've tested it can be parsed normally
<img width="1695" alt="image" src="https://github.com/paritytech/substrate-api-sidecar/assets/111047493/b3bc4847-2b03-467b-95b4-a07d791ea8b9">

**Comment by web3olalala** [comment]
And I also noticed that `https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fbifrost-parachain.api.onfinality.io%2Fpublic-ws#/explorer/query/2652989` .
 Polkadot.js is able to parse this block properly, you can also see what they are parsing

**Comment by hqwangningbo** [comment]
@TarikGul Because the key is an Object type, so the error is reported, my side to deal with, and then it can be parsed normally, I do not know this solution is feasible
<img width="1856" alt="image" src="https://github.com/paritytech/substrate-api-sidecar/assets/57781136/c4a067b1-faf4-4ecc-8d46-a2cb44e9fd8f">

**Comment by hqwangningbo** [comment]
And I also noticed that `https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fbifrost-parachain.api.onfinality.io%2Fpublic-ws#/explorer/query/2652989`. Polkadot.js is able to parse this block properly, you can also see what they are parsing

**Comment by Imod7** [comment]
I noticed that connecting Sidecar to different Bifrost endpoint results in a valid response and not an error. Mentioning it here in case it is helpful.

If I connect Sidecar to one of Bifrost's [public endpoints](https://github.com/bifrost-platform/bifrost-node#bifrost-network-mainnet-chainid-3068) :
- wss://public-01.mainnet.bifrostnetwork.com/wss
- wss://public-02.mainnet.bifrostnetwork.com/wss

and then I request `http://127.0.0.1:8080/blocks/2652989`, I do not get a `500` error but instead I get a valid response :

```
{
  "number": "2652989",
  "hash": "0xded64732db2c3f3bace7f4a29ee63c0ada4a7cb66b6c045305089612ddcdae38",
  "parentHash": "0x67211a5679241b25bc880b4591f5aad0b9c6652b818bb1c6710de1296f6a1720",
  "stateRoot": "0xc255b0b4f280c3a261f2b4240d70470dcffdb86671fc47f54de5e2915037c652",
  "extrinsicsRoot": "0x61ecef9d6e1d5fc232ba19cd0d5661e513ead796a5482d296d9a1564e4f3269f",
  "authorId": "0x6bEf93e6d6bC1e02B9D697b4FB8606152C200b29",
  "logs": [
    {
      "type": "PreRuntime",
      "index": "6",
      "value": [
        "0x61757261",
        "0x11cc682100000000"
      ]
    },
    {
      "type": "Consensus",
....
....
```

However, as soon as I use `wss://bifrost-rpc.liebi.com/ws` endpoint mentioned in the [description](https://github.com/paritytech/substrate-api-sidecar/issues/1082#issue-1405460385), I get the error message.

**Comment by Imod7** [comment]
Closing this issue since the [fix](https://github.com/paritytech/substrate-api-sidecar/pull/1345) that was included in the release [v17.3.1](https://github.com/paritytech/substrate-api-sidecar/releases/tag/v17.3.1) should resolve any errors and show the blocks information correctly. If this error appears again in the future, please reopen this issue.
